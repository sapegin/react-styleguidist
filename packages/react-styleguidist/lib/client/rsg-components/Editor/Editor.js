import "core-js/modules/es7.symbol.async-iterator";
import "core-js/modules/es6.symbol";
import "core-js/modules/web.dom.iterable";
import "core-js/modules/es6.array.iterator";
import "core-js/modules/es6.object.keys";
import "core-js/modules/es6.string.link";
import "core-js/modules/es6.string.small";

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Styled from 'rsg-components/Styled';
import { polyfill } from 'react-lifecycles-compat';
import SimpleEditor from 'react-simple-code-editor';
import Prism from 'prismjs/components/prism-core';
import 'prismjs/components/prism-clike';
import 'prismjs/components/prism-markup';
import 'prismjs/components/prism-javascript';
import 'prismjs/components/prism-jsx';
import { space } from '../../styles/theme';
import prismTheme from '../../styles/prismTheme';

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (let i = 0; i < props.length; i++) { const descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) {descriptor.writable = true;} Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) {_defineProperties(Constructor.prototype, protoProps);} if (staticProps) {_defineProperties(Constructor, staticProps);} return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) {_setPrototypeOf(subClass, superClass);} }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _objectSpread(target) { for (let i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; let ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const highlight = function highlight(code) {
  return Prism.highlight(code, Prism.languages.jsx, 'jsx');
};

const styles = function styles(_ref) {
  const fontFamily = _ref.fontFamily;

      
const fontSize = _ref.fontSize;

      
const color = _ref.color;

      
const borderRadius = _ref.borderRadius;
  return {
    root: _objectSpread({
      fontFamily: fontFamily.monospace,
      fontSize: fontSize.small,
      background: color.codeBackground,
      borderRadius,
      '& textarea': {
        isolate: false,
        transition: 'all ease-in-out .1s',
        // important to override inline styles in react-simple-code-editor
        border: "1px ".concat(color.border, " solid !important"),
        borderRadius
      },
      '& textarea:focus': {
        isolate: false,
        outline: 0,
        borderColor: "".concat(color.link, " !important"),
        boxShadow: [[0, 0, 0, 2, color.focus]]
      }
    }, prismTheme({
      color
    }))
  };
};

export var Editor =
/*#__PURE__*/
function (_Component) {
  _inherits(Editor, _Component);

  function Editor() {
    let _getPrototypeOf2;

    let _this;

    _classCallCheck(this, Editor);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(Editor)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "state", {
      code: _this.props.code,
      prevCode: _this.props.code
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "handleChange", function (code) {
      _this.setState({
        code
      });

      _this.props.onChange(code);
    });

    return _this;
  }

  _createClass(Editor, [{
    key: "shouldComponentUpdate",
    value: function shouldComponentUpdate(nextProps, nextState) {
      return nextState.code !== this.state.code;
    }
  }, {
    key: "render",
    value: function render() {
      return React.createElement(SimpleEditor, {
        className: this.props.classes.root,
        value: this.state.code,
        onValueChange: this.handleChange,
        highlight // Padding should be passed via a prop (not CSS) for a proper
        // cursor position calculation
        ,
        padding: space[2]
      });
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, prevState) {
      const code = nextProps.code;

      if (prevState.prevCode !== code) {
        return {
          prevCode: code,
          code
        };
      }

      return null;
    }
  }]);

  return Editor;
}(Component);

_defineProperty(Editor, "propTypes", {
  code: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  classes: PropTypes.object.isRequired
});

export default Styled(styles)(polyfill(Editor));