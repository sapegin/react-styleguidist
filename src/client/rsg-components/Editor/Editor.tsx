import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Styled, { JssInjectedProps } from 'rsg-components/Styled';
import SimpleEditor from 'react-simple-code-editor';
import MdxHighlight from 'rsg-components/mdx/MdxHighlight';
import * as Rsg from '../../../typings';

const highlight = (code: string) => <MdxHighlight className="language-jsx">{code}</MdxHighlight>;

const styles = ({ space, fontFamily, fontSize, color, borderRadius }: Rsg.Theme) => ({
	root: {
		// Use `important` to override inline styles in react-simple-code-editor
		padding: `${space[2]}px !important`,
		fontFamily: fontFamily.monospace,
		fontSize: fontSize.small,
		background: color.codeBackground,
		borderRadius,
		lineHeight: 1.5,
		tabSize: 2,
		'& textarea': {
			padding: 'inherit !important',
			isolate: false,
			transition: 'all ease-in-out .1s',
			border: `1px ${color.border} solid !important`,
			borderRadius,
		},
		'& textarea:focus': {
			isolate: false,
			outline: 0,
			borderColor: `${color.link} !important`,
			boxShadow: [[0, 0, 0, 2, color.focus]],
		},
	},
});

export interface EditorProps extends JssInjectedProps {
	code: string;
	onChange: (code: string) => void;
}

interface EditorState {
	code: string;
	prevCode: string;
}

export class Editor extends Component<EditorProps> {
	public static propTypes = {
		code: PropTypes.string.isRequired,
		onChange: PropTypes.func.isRequired,
		classes: PropTypes.objectOf(PropTypes.string.isRequired).isRequired,
	};

	public state = { code: this.props.code, prevCode: this.props.code };

	public static getDerivedStateFromProps(nextProps: EditorProps, prevState: EditorState) {
		const { code } = nextProps;
		if (prevState.prevCode !== code) {
			return {
				prevCode: code,
				code,
			};
		}
		return null;
	}

	public shouldComponentUpdate(nextProps: EditorProps, nextState: EditorState) {
		return nextState.code !== this.state.code;
	}

	private handleChange = (code: string) => {
		this.setState({ code });
		this.props.onChange(code);
	};

	public render() {
		return (
			<SimpleEditor
				className={this.props.classes.root}
				value={this.state.code}
				onValueChange={this.handleChange}
				highlight={highlight}
			/>
		);
	}
}

export default Styled<EditorProps>(styles)(Editor);
