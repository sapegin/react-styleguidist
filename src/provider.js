import React, { Component } from 'react';
import PropTypes from 'prop-types';

export const Context = React.createContext();

export function Consumer(WrappedComponent) {
	const componentName = WrappedComponent.name.replace(/Renderer$/, '');
	return class extends Component {
		static displayName = `Consumer(${componentName})`;
		state = {};
		render() {
			return (
				<Context.Consumer>
					{context => {
						if (!context) {
							throw new Error(`Consumer cannot be rendered outside the Provider component`);
						}
						return <WrappedComponent {...context} {...this.props} />;
					}}
				</Context.Consumer>
			);
		}
	};
}

export function Provider(props) {
	const { children, ...rest } = props;
	const ui = typeof children === 'function' ? children(this.state) : children;

	return <Context.Provider value={{ ...rest }}>{ui}</Context.Provider>;
}

Provider.propTypes = {
	children: PropTypes.any,
	codeRevision: PropTypes.number.isRequired,
	config: PropTypes.object.isRequired,
	slots: PropTypes.object.isRequired,
	displayMode: PropTypes.string,
};
