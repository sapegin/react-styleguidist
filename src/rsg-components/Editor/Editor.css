/* Tweak CodeMirror stylesÂ */
.root :global .CodeMirror,
.root_err :global .CodeMirror,
.root_edit :global .CodeMirror {
	height: auto;
	padding: 5px 12px;
	font-size: 12px;
}
.root_err {
	border-left: red 4px solid;
}
.root_edit {
	border-left: #1978c8 4px solid;
}
.root {
	border-left: rgb(245, 245, 245) 4px solid;
}
.root :global .CodeMirror-scroll,
.root_err :global .CodeMirror-scroll,
.root_edit :global .CodeMirror-scroll
 {
	height: auto;
	overflow-y: hidden;
	overflow-x: auto;
}
.root :global .cm-error,
.root_err :global .cm-error,
.root_edit :global .cm-error
 {
	background: none !important;
}
