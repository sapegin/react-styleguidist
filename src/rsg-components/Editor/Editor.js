import React, { Component } from 'react';
import PropTypes from 'prop-types';
import debounce from 'lodash/debounce';
import CodeMirror from 'react-codemirror2';
import 'codemirror/mode/jsx/jsx';

// Weâ€™re explicitly specifying Webpack loaders here so we could skip specifying them in Webpack configuration.
// That way we could avoid clashes between our loaders and user loaders.
// eslint-disable-next-line import/no-unresolved
require('!!../../../loaders/style-loader!../../../loaders/css-loader!codemirror/lib/codemirror.css');
// eslint-disable-next-line import/no-unresolved
require('!!../../../loaders/style-loader!../../../loaders/css-loader!rsg-codemirror-theme.css');

const codemirrorOptions = {
	mode: 'jsx',
	lineNumbers: false,
	lineWrapping: true,
	smartIndent: false,
	matchBrackets: true,
	viewportMargin: Infinity,
};

const UPDATE_DELAY = 10;

export default class Editor extends Component {
	static propTypes = {
		code: PropTypes.string.isRequired,
		onChange: PropTypes.func.isRequired,
	};
	static contextTypes = {
		config: PropTypes.object.isRequired,
	};

	constructor() {
		super();
		this.handleChange = debounce(this.handleChange.bind(this), UPDATE_DELAY);
	}

	shouldComponentUpdate() {
		return false;
	}

	handleChange(editor, metadata, newCode) {
		this.props.onChange(newCode);
	}

	render() {
		const { code } = this.props;
		const { highlightTheme } = this.context.config;
		const options = {
			...codemirrorOptions,
			theme: highlightTheme,
		};
		return <CodeMirror value={code} onValueChange={this.handleChange} options={options} />;
	}
}
