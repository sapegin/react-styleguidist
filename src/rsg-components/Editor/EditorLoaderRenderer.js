import React from 'react';
import PropTypes from 'prop-types';
import Styled from 'rsg-components/Styled';

const styles = ({ font, monospace, light, codeBackground }) => ({
	root: {
		padding: [[7, 16, 10, 7]],
		fontFamily: font,
		fontSize: 12,
		color: light,
		backgroundColor: codeBackground,
	},
	// Tweak CodeMirror styles
	'@global': {
		body: {
			'& .CodeMirror': {
				fontFamily: monospace,
				height: 'auto',
				padding: [[5, 12]],
				fontSize: 12,
			},
			'& .CodeMirror-scroll': {
				height: 'auto',
				overflowY: 'hidden',
				overflowX: 'auto',
			},
			'& .cm-error': {
				background: 'none !important',
			},
		},
	},
});

export function EditorLoaderRenderer({ classes }) {
	return (
		<div className={classes.root}>
			Loadingâ€¦
		</div>
	);
}

EditorLoaderRenderer.propTypes = {
	classes: PropTypes.object.isRequired,
};

export default Styled(styles)(EditorLoaderRenderer);
