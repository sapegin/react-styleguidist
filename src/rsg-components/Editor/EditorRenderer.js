import React, { PropTypes } from 'react';

import s from './Editor.css';

const EditorRenderer = ({ children, isValid, isTyping }) => {
	const typingStyle = isTyping ? s.root_edit : s.root;
	return (
		<div className={isValid ? typingStyle : s.root_err}>
			{children}
		</div>);
};

EditorRenderer.propTypes = {
	children: PropTypes.node.isRequired,
	isValid: PropTypes.bool.isRequired,
	isTyping: PropTypes.bool.isRequired,
};

export default EditorRenderer;
