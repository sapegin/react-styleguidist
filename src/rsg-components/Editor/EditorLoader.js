import React, { Component } from 'react';
import EditorLoaderRenderer from 'rsg-components/Editor/EditorLoaderRenderer';
import Editor from './Editor';

export default class EditorLoader extends Component {
	state = {
		editor: null,
	};

	// componentDidMount() {
   //  System.import('rsg-components/Editor/Editor').then(module => {
   //    console.log(module);
   //    this.setState({ editor: module.default });
   //  }).catch((err) => {
   //    console.log(err);
   //  });
	// }

	render() {
		if (Editor) {
			return <Editor {...this.props} />;
		}

		return <EditorLoaderRenderer />;
	}
}
