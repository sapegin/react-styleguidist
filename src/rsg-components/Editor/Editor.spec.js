import React from 'react';
import { EditorLoaderRenderer } from './EditorLoaderRenderer';
import { Editor } from './Editor';
import { Provider } from '../../provider';

const code = '<button>MyAwesomeCode</button>';
const newCode = '<button>MyNewAwesomeCode</button>';
const context = {
	config: {
		showCode: false,
		highlightTheme: 'base16-light',
		editorConfig: {
			lineWrapping: true,
			mode: 'js',
		},
	},
};

describe('EditorLoaderRenderer', () => {
	it('should renderer should render loader', () => {
		const actual = shallow(<EditorLoaderRenderer classes={{}} />);

		expect(actual).toMatchSnapshot();
	});
});

describe('Editor', () => {
	it('should renderer and editor', () => {
		const actual = shallow(
			<Provider {...context}>
				<Editor classes={{}} code={code} editorConfig={{ readOnly: true, mode: 'text/html' }} />,
			</Provider>
		);

		expect(actual).toMatchSnapshot();
	});

	it('should update code with debounce', done => {
		const onChange = jest.fn();
		const actual = mount(
			<Provider {...context}>
				<Editor classes={{}} code={code} onChange={onChange} />
			</Provider>
		);

		expect(actual.text()).toMatch(code);

		// Set new value by calling a method on the CodeMirror instance
		actual
			.find('div')
			.getDOMNode()
			.querySelector('.CodeMirror')
			.CodeMirror.setValue(newCode);

		setTimeout(() => {
			expect(onChange).toBeCalledWith(newCode);
			done();
		}, 13);
	});

	it('should not update if not read only', () => {
		const actual = shallow(
			<Provider {...context}>
				<Editor classes={{}} code={code} />
			</Provider>
		);

		const shouldUpdate = actual.instance().shouldComponentUpdate({ code: newCode });
		expect(shouldUpdate).toBe(false);
	});

	it('should update if read only and code has changed', () => {
		const actual = shallow(
			<Provider {...context}>
				<Editor classes={{}} code={code} editorConfig={{ readOnly: true }} />
			</Provider>
		);

		const shouldUpdate = actual
			.instance()
			.shouldComponentUpdate({ code: newCode, editorConfig: { readOnly: true } });
		expect(shouldUpdate).toBe(true);
	});
});
