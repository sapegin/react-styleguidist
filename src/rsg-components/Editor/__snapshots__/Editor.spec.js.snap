// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor should renderer and editor 1`] = `
<UnControlled
  onChange={[Function]}
  options={
    Object {
      "lineNumbers": false,
      "lineWrapping": true,
      "matchBrackets": true,
      "mode": "jsx",
      "smartIndent": false,
      "theme": "base16-light",
      "viewportMargin": Infinity,
    }
  }
  value="<button>MyAwesomeCode</button>"
/>
`;

exports[`EditorLoaderRenderer should renderer should render loader 1`] = `
<div>
  Loadingâ€¦
</div>
`;
